lib/iinChecker.js

```diff
@@ -132,18 +73,178 @@
+          emitter.emit('requestFailed', iinToLookup, _request);
+          return false;
+        }
+        // Remap the reply to match our schema
+        cardInfo = provider.map( body, self.nullValue );
+        // we have some issues with binlist's correctness, some of their binranges
+        // starting with 5 are being returned as VISA when they are actually MASTERCARD
+        cardInfo = self.flipBrandWhereNecessary( cardInfo );
+        // if caching is on, provider request is only attempted if card details were not
+        // in the cache now we have the details set it.
+        if ( self.isCacheable( cardInfo ) ) {
+          self.options.cache.set( iinToLookup, cardInfo );
+        }
+        // fire off an event so we know that we have completed our request
+        emitter.emit('requestComplete', cardInfo);
+        return true;
+      });
+
+      // fire off an event so we know that the request has started
+      emitter.emit('requestStarted', _request, iinToLookup);
+    }
+
+    // Loop through the providers and make a request to each of them
+    function doLookup( providersToCheck, iinToCheck ) {
+      providersToCheck.forEach(function( element ) {
+        doRequest( iinToCheck, element );
+      });
+    }
+
+    // Do a RegEx lookup on a iin
+    function doRegex( iinToLookup ) {
+      var cardInfo;
+      var patterns = require( '../configs/patterns' );
+
+      // Set out null info object
+      cardInfo = {
+        iin: iinToLookup,
+        brand: self.nullValue,
+        issuer: self.nullValue,
+        type: self.nullValue,
+        category: self.nullValue,
+        country: self.nullValue
+      };
+
+      _.each( patterns, function( pattern ) {
+        var regex = new RegExp( pattern.expression );
+        if ( cardInfo.brand === self.nullValue && regex.test( iinToLookup ) ) {
+          cardInfo.brand = self.brands[pattern.name];
+        }
+      } );
+
+      // See if RegEx worked
+      if ( cardInfo.brand && cardInfo.brand !== self.brands.UNKNOWN ) {
+        // Our RegEx has work, so we fire off an event so we know about it
+        emitter.emit('regexResult', cardInfo);
+      } else {
+        // The RegEx was unable to figure out what the card was so we will fire off an event so we know about it and deal with it
+        emitter.emit('returnError', new TypeError(self.options.messages.PARAMETER_IIN_IS_INVALID), iinToLookup);
+      }
+    }
+
+    // Bind our listeners here so we can handle events
+    function bindListeners( listenerCallback ) {
+      // we need to keep track of any requests we have
+      var currentRequests = [];
+
+      function stopListening() {
+        emitter.removeAllListeners('requestStarted')
+          .removeAllListeners('requestComplete')
+          .removeAllListeners('requestFailed')
+          .removeAllListeners('regexResult')
+          .removeAllListeners('returnError');
+      }
+
+      function clearRequestQueue() {
```

Comment:
i've never used it before but would `_.invokeMap(currentRequests, 'abort');` work here?